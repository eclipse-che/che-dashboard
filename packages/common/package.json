{
  "name": "@eclipse-che/common",
  "version": "7.34.0-SNAPSHOT",
  "repository": "https://github.com/eclipse-che/che-dashboard",
  "license": "EPL-2.0",
  "private": true,
  "main": "lib/index.js",
  "engines": {
    "yarn": ">=1.20.0",
    "node": ">=12"
  },
  "scripts": {
    "build": "yarn clean && tsc -b",
    "watch": "yarn build && tsc -w",
    "clean": "rimraf lib tsconfig.tsbuildinfo"
  },
  "devDependencies": {
    "axios": "^0.21.1",
    "rimraf": "^3.0.2",
    "typescript": "~4.4.2"
  }
}
