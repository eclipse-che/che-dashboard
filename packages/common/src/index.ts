/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import helpers from './helpers';
import * as api from './dto/api';

export * from './dto/cluster-info';
export * from './dto/cluster-config';

export { helpers, api };

export const FACTORY_LINK_ATTR = 'factoryLink';
export const ERROR_CODE_ATTR = 'error_code';
export const KUBECONFIG_MOUNT_PATH = '/tmp/.kube';

const common = {
  helpers,
  api,
};
export default common;
