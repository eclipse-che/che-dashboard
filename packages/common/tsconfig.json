{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "composite": true
  },
  "include": [
    "src",
  ],
  "exclude": [
    "node_modules",
    "lib",
  ]
}
