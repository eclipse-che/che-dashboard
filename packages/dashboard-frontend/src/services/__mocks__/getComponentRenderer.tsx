/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
