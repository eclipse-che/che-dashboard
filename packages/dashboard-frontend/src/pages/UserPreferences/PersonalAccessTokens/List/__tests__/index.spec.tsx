/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { api } from '@eclipse-che/common';
import React from 'react';

import getComponentRenderer, {
  fireEvent,
  screen,
  within,
} from '@/services/__mocks__/getComponentRenderer';

import { PersonalAccessTokenList } from '..';

jest.mock('../Toolbar');

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnAddToken = jest.fn();
const mockOnEditToken = jest.fn();
const mockOnDeleteToken = jest.fn();

describe('PersonalAccessTokenList', () => {
  let tokens: api.PersonalAccessToken[];

  beforeEach(() => {
    tokens = [
      {
        cheUserId: 'che-user',
        gitProvider: 'github',
        gitProviderEndpoint: 'https://github.com',
        tokenData: 'token-data-1',
        tokenName: 'token-name-1',
      } as api.PersonalAccessToken,
      {
        cheUserId: 'che-user',
        gitProvider: 'azure-devops',
        gitProviderEndpoint: 'https://dev.azure.com',
        gitProviderOrganization: 'dev-azure-org',
        tokenData: 'token-data-2',
        tokenName: 'token-name-2',
      } as api.PersonalAccessToken,
    ];
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const snapshot = createSnapshot(tokens);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  describe('list head row', () => {
    test('disable head checkbox', () => {
      renderComponent(tokens, true);
      const selectAllCheckbox = screen.getByRole('checkbox', {
        name: 'Select all rows',
      });

      expect(selectAllCheckbox).toBeDisabled();
    });

    test('select all tokens', () => {
      renderComponent(tokens);
      const selectAllCheckbox = screen.getByRole('checkbox', {
        name: 'Select all rows',
      });
      const token1Checkbox = screen.getByRole('checkbox', { name: 'Select row 0' });
      const token2Checkbox = screen.getByRole('checkbox', { name: 'Select row 1' });

      // select all tokens
      fireEvent.click(selectAllCheckbox);

      expect(selectAllCheckbox).toBeChecked();
      expect(token1Checkbox).toBeChecked();
      expect(token2Checkbox).toBeChecked();

      // deselect all tokens
      fireEvent.click(selectAllCheckbox);

      expect(selectAllCheckbox).not.toBeChecked();
      expect(token1Checkbox).not.toBeChecked();
      expect(token2Checkbox).not.toBeChecked();
    });
  });

  describe('list body row', () => {
    test('disable row checkbox', () => {
      renderComponent(tokens, true);
      const selectRowCheckbox = screen.getByRole('checkbox', {
        name: 'Select row 0',
      });

      expect(selectRowCheckbox).toBeDisabled();
    });

    test('select one token', () => {
      renderComponent(tokens);
      const selectAllCheckbox = screen.getByRole('checkbox', {
        name: 'Select all rows',
      });
      const token1Checkbox = screen.getByRole('checkbox', { name: 'Select row 0' });
      const token2Checkbox = screen.getByRole('checkbox', { name: 'Select row 1' });

      // select token 1
      fireEvent.click(token1Checkbox);

      expect(selectAllCheckbox).not.toBeChecked();
      expect(token1Checkbox).toBeChecked();
      expect(token2Checkbox).not.toBeChecked();

      // select token 2
      fireEvent.click(token2Checkbox);

      expect(selectAllCheckbox).toBeChecked();
      expect(token1Checkbox).toBeChecked();
      expect(token2Checkbox).toBeChecked();

      // deselect token 1
      fireEvent.click(token1Checkbox);

      expect(selectAllCheckbox).not.toBeChecked();
      expect(token1Checkbox).not.toBeChecked();
      expect(token2Checkbox).toBeChecked();
    });
  });

  describe('toolbar buttons', () => {
    test('disable buttons', () => {
      renderComponent(tokens, true);

      const addButton = screen.getByRole('button', { name: 'Add Token' });
      const deleteButton = screen.getByRole('button', { name: 'Delete' });

      expect(addButton).toBeDisabled();
      expect(deleteButton).toBeDisabled();
    });

    test('add a new token', () => {
      renderComponent(tokens);

      expect(mockOnAddToken).not.toHaveBeenCalled();

      const button = screen.getByRole('button', { name: 'Add Token' });
      fireEvent.click(button);

      expect(mockOnAddToken).toHaveBeenCalled();
    });

    test('delete selected tokens', () => {
      renderComponent(tokens);

      expect(mockOnDeleteToken).not.toHaveBeenCalled();

      const token1Row = screen.getByRole('row', { name: new RegExp(tokens[0].tokenName) });
      const token1Checkbox = within(token1Row).getByRole('checkbox', { name: 'Select row 0' });

      const deleteButton = screen.getByRole('button', { name: 'Delete' });

      // select token 1
      fireEvent.click(token1Checkbox);

      // delete selected tokens
      fireEvent.click(deleteButton);

      expect(mockOnDeleteToken).toHaveBeenCalledWith([tokens[0]]);
    });
  });

  describe('row actions', () => {
    test('disable actions menu', () => {
      renderComponent(tokens, true);

      const token1Row = screen.getByRole('row', { name: new RegExp(tokens[0].tokenName) });
      const actionsButton = within(token1Row).getByRole('button', { name: 'Actions' });

      expect(actionsButton).toBeDisabled();
    });

    test('edit token', () => {
      renderComponent(tokens);

      expect(mockOnEditToken).not.toHaveBeenCalled();

      const token1Row = screen.getByRole('row', { name: new RegExp(tokens[0].tokenName) });
      const actionsButton = within(token1Row).getByRole('button', { name: 'Actions' });

      // open actions menu
      fireEvent.click(actionsButton);

      const token1EditButton = screen.getByRole('menuitem', { name: 'Edit Token' });

      // edit token 1
      fireEvent.click(token1EditButton);

      expect(mockOnEditToken).toHaveBeenCalledWith(tokens[0]);
    });

    test('delete token', async () => {
      renderComponent(tokens);

      expect(mockOnDeleteToken).not.toHaveBeenCalled();

      const token1Row = screen.getByRole('row', { name: new RegExp(tokens[0].tokenName) });
      const actionsButton = within(token1Row).getByRole('button', { name: 'Actions' });

      // open actions menu
      fireEvent.click(actionsButton);

      const token1DeleteButton = screen.getByRole('menuitem', { name: 'Delete Token' });

      // delete token 1
      fireEvent.click(token1DeleteButton);

      expect(mockOnDeleteToken).toHaveBeenCalledWith([tokens[0]]);
    });
  });
});

function getComponent(tokens: api.PersonalAccessToken[], isDisabled = false): React.ReactElement {
  return (
    <PersonalAccessTokenList
      isDisabled={isDisabled}
      tokens={tokens}
      onAddToken={mockOnAddToken}
      onEditToken={mockOnEditToken}
      onDeleteTokens={mockOnDeleteToken}
    />
  );
}
