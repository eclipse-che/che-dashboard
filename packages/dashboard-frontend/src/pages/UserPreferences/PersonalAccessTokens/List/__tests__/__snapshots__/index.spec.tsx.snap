// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalAccessTokenList should match the snapshot 1`] = `
<section
  className="pf-c-page__main-section"
>
  <div
    data-testid="personal-access-token-list-toolbar"
  >
    <button
      disabled={false}
      onClick={[Function]}
    >
      Add Token
    </button>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
  <table
    aria-label="Personal Access Tokens List"
    className="pf-c-table pf-m-grid-md pf-m-compact"
    data-ouia-component-id="OUIA-Generated-Table-1"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe={true}
    role="grid"
  >
    <thead
      className=""
    >
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-1"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <td
          className="pf-c-table__check"
          onMouseEnter={[Function]}
          scope={null}
        >
          <label>
            <input
              aria-label="Select all rows"
              checked={false}
              name="check-all"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <th
          className=""
          onMouseEnter={[Function]}
          scope="col"
        >
          Name
        </th>
        <th
          className=""
          onMouseEnter={[Function]}
          scope="col"
        >
          Provider
        </th>
        <th
          className=""
          onMouseEnter={[Function]}
          scope="col"
        >
          Endpoint
        </th>
        <td
          className=""
          onMouseEnter={[Function]}
        />
      </tr>
    </thead>
    <tbody
      className=""
      role="rowgroup"
    >
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-2"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <td
          className="pf-c-table__check"
          onMouseEnter={[Function]}
        >
          <label>
            <input
              aria-label="Select row 0"
              checked={false}
              name="checkrow0"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <td
          className=""
          data-label="Name"
          onMouseEnter={[Function]}
        >
          token-name-1
        </td>
        <td
          className=""
          data-label="Provider"
          onMouseEnter={[Function]}
        >
          github
        </td>
        <td
          className=""
          data-label="Endpoint"
          onMouseEnter={[Function]}
        >
          https://github.com
        </td>
        <td
          className="pf-c-table__action"
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-1"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="Actions"
              className="pf-c-dropdown__toggle pf-m-plain"
              disabled={false}
              id="pf-dropdown-toggle-id-0"
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-3"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <td
          className="pf-c-table__check"
          onMouseEnter={[Function]}
        >
          <label>
            <input
              aria-label="Select row 1"
              checked={false}
              name="checkrow1"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <td
          className=""
          data-label="Name"
          onMouseEnter={[Function]}
        >
          token-name-2
        </td>
        <td
          className=""
          data-label="Provider"
          onMouseEnter={[Function]}
        >
          azure-devops
        </td>
        <td
          className=""
          data-label="Endpoint"
          onMouseEnter={[Function]}
        >
          https://dev.azure.com
        </td>
        <td
          className="pf-c-table__action"
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-2"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="Actions"
              className="pf-c-dropdown__toggle pf-m-plain"
              disabled={false}
              id="pf-dropdown-toggle-id-1"
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;
